core/src/main/java/org/infinispan/transaction/tm/DummyTransaction.java

```diff
@@ -164,160 +231,81 @@ public boolean delistResource(XAResource xaRes, int flag)
-         }
-      }
-      return retval;
+   public Collection<XAResource> getEnlistedResources() {
+      return Collections.unmodifiableList(resources);
    }
 
-   public boolean runPrepare() throws SystemException {
-
-      boolean successfulInit = notifyBeforeCompletion();
+   public boolean runPrepare() {
+      if (trace) {
+         log.tracef("runPrepare() invoked in transaction with Xid=%s", xid);
+      }
+      notifyBeforeCompletion();
+      endResources();
 
-      if (successfulInit) {
-         status = Status.STATUS_PREPARING;
-      } else {
-         status = Status.STATUS_ROLLING_BACK;
+      if (status == Status.STATUS_MARKED_ROLLBACK) {
+         return false;
       }
 
-      DummyTransaction transaction = tm_.getTransaction();
-      Collection<XAResource> resources = transaction.getEnlistedResources();
-      for (XAResource res : resources) {
+      status = Status.STATUS_PREPARING;
+
+      for (XAResource res : getEnlistedResources()) {
+         //note: it is safe to return even if we don't prepare all the resources. rollback will be invoked.
          try {
-            int prepareStatus = res.prepare(xid);
-            transaction.setPrepareStatus(prepareStatus);
+            if (trace) {
+               log.tracef("XaResource.prepare() for %s", res);
+            }
+            //don't need to check return value. the only possible values are OK or READ_ONLY.
+            res.prepare(xid);
          } catch (XAException e) {
-            log.trace("The resource wants to rollback!", e);
-            status = Status.STATUS_ROLLING_BACK;
+            if (trace) {
+               log.trace("The resource wants to rollback!", e);
+            }
+            markRollbackOnly(newRollbackException(format("XaResource.prepare() for %s wants to rollback.", res), e));
             return false;
          } catch (Throwable th) {
+            markRollbackOnly(newRollbackException(format("Unexpected error in XaResource.prepare() for %s. Rollback transaction.", res), th));
             log.unexpectedErrorFromResourceManager(th);
-            throw new SystemException(th.getMessage());
+            return false;
          }
       }
-
-      if (status == Status.STATUS_MARKED_ROLLBACK || status == Status.STATUS_ROLLING_BACK) {
-         return false;
-      }
-
       status = Status.STATUS_PREPARED;
       return true;
    }
 
-   private void setPrepareStatus(int prepareStatus) {
-      this.prepareStatus = prepareStatus;
-   }
-
-   public void notifyAfterCompletion(int status) {
-      if (syncs == null) return;
-      for (Synchronization s : syncs) {
-         if (trace) {
-            log.tracef("processing afterCompletion for %s", s);
-         }
-         try {
-            s.afterCompletion(status);
-         } catch (Throwable t) {
-            log.afterCompletionFailed(s, t);
-         }
+   public void runCommit(boolean forceRollback) throws HeuristicMixedException, HeuristicRollbackException {
```

Comment:
for 2 reasons: 
1) the code is very similar. only the status changes and the method invoked in the XaResource
2) want to avoid errors. If the runPrepare() fails I don't wanna invoke the runCommitTx() instead of runRollbackTx(). This way, you can force a transaction to rollback but never to commit.